torch/utils/data/datapipes/iter/combining.py

```diff
@@ -151,32 +151,40 @@ def get_next_element_by_instance(self, instance_id: int):
-            # Use buffer
-            if self.buffer and self.child_pointers[instance_id] <= self.leading_ptr:
-                idx = self.child_pointers[instance_id] - self.slowest_ptr - 1
-                return_val = self.buffer[idx]
-            else:  # Retreive one element from main datapipe
-                self.leading_ptr = self.child_pointers[instance_id]
-                try:
-                    return_val = next(self._datapipe_iterator)
-                    self.buffer.append(return_val)
-                except StopIteration:
-                    self.end_ptr = self.leading_ptr
-                    continue
-            if self.child_pointers[instance_id] == self.slowest_ptr + 1:
-                new_min = min(self.child_pointers)  # Can optimize by avoiding the call to min()
-                if self.slowest_ptr < new_min:
-                    self.slowest_ptr = new_min
-                    self.buffer.popleft()
-            if self.buffer_size >= 0 and self.leading_ptr > self.buffer_size + self.slowest_ptr:
-                raise BufferError("ForkerIterDataPipe buffer overflow," +
-                                  f"buffer size {self.buffer_size} is insufficient.")
-            yield return_val
-
-        if all(p + 1 == self.end_ptr for p in self.child_pointers):
+        try:
+            while self.end_ptr is None or self.child_pointers[instance_id] + 1 < self.end_ptr:
+                self.child_pointers[instance_id] += 1
+                # Use buffer
+                if self.buffer and self.child_pointers[instance_id] <= self.leading_ptr:
+                    idx = self.child_pointers[instance_id] - self.slowest_ptr - 1
+                    return_val = self.buffer[idx]
+                else:  # Retreive one element from main datapipe
+                    self.leading_ptr = self.child_pointers[instance_id]
+                    try:
+                        return_val = next(self._datapipe_iterator)
+                        self.buffer.append(return_val)
+                    except StopIteration:
+                        self.end_ptr = self.leading_ptr
+                        continue
+                if self.child_pointers[instance_id] == self.slowest_ptr + 1:
+                    new_min = min(self.child_pointers)  # Can optimize by avoiding the call to min()
+                    if self.slowest_ptr < new_min:
+                        self.slowest_ptr = new_min
+                        self.buffer.popleft()
+                if self.buffer_size >= 0 and self.leading_ptr > self.buffer_size + self.slowest_ptr:
+                    raise BufferError("ForkerIterDataPipe buffer overflow," +
+                                      f"buffer size {self.buffer_size} is insufficient.")
+                yield return_val
+
+            if all(p + 1 == self.end_ptr for p in self.child_pointers):
+                self._datapipe_iterator = None
+        finally:
             self._datapipe_iterator = None
+            if self.buffer:
```

Comment:
curious: why we cannot wait for `self.__del__` being triggered? (looks like the clean up is the same).

Or, does that makes sense to refactor some `_cleanup` and make both `__del__` and here call it? ~

Reply:
> curious: why we cannot wait for `self.__del__` being triggered? (looks like the clean up is the same).

The main problem is that `__del__` is not always executed in Python.


> does that makes sense to refactor some `_cleanup` and make both `__del__` and here call it? ~

Yeah, it makes sense to me.

Reply:
Agree that we can refactor to `_cleanup` but I'm also okay with this as it is since it is only called twice